torch/testing/_internal/common_distributed.py

```diff
@@ -834,3 +839,69 @@ def tp_transports():
     see https://github.com/pytorch/pytorch/issues/73885 and https://github.com/pytorch/pytorch/issues/65022
     """
     return ["shm", "uv"] if has_efa() else None
+
+
+
+def spawn_threads_and_init_comms(func=None, timeout=TIMEOUT_DEFAULT, world_size=DEFAULT_WORLD_SIZE):
+    """
+    Wrapper to use with a test method
+    """
+    if func is None:
+        return partial(spawn_threads_and_init_comms, timeout=timeout, world_size=world_size)
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        failed_ranks = run_with_threaded_pg(world_size, timeout, lambda: func(self, *args, **kwargs))
+        for rank, exc_info in failed_ranks:
+            print(f"Rank {rank} raised:")
+            for line in traceback.format_exception(*exc_info):
+                sys.stdout.write(line)
+        self.assertEqual([], failed_ranks, "Some ranks failed")
+
+
+    return wrapper
+
+class MultiThreadedTestCase(TestCase):
+    """
+    Simple test runner that executes all tests with the in-proc process group.
+
+    A single instance of the TestCase object for all threads.
+
+    Difference from regular test runner:
+    Cannot use setUp / tearDown (must use perThreadSetup / perThreadShutdown)
+        Not sure what these two would be good for though.
+    No global state possible
+        How bad of a limitation is this?
+    """
+
+    def __init__(self, method_name: str = "runTest") -> None:
+        super().__init__(method_name)
+        self._test_method = getattr(self, method_name, None)
+        setattr(self, method_name, self.threaded_run_test)
+        if TestCase.setUp != type(self).setUp:
+            raise RuntimeError(f"Test class {type(self)} overrides disabled method setUp. Use perThreadSetUp instead")
+        if TestCase.tearDown != type(self).tearDown:
+            raise RuntimeError(f"Test class {type(self)} overrides disabled method tearDown. Use perThreadTearDown instead")
+
+
+    def threaded_run_test(self):
+        self.perThreadSetUp()
+        try:
+            failed_ranks = run_with_threaded_pg(self.world_size, TIMEOUT_DEFAULT, self._test_method)
```

Comment:
code looks similar to `spawn_threads_and_init_comms`, wonder why we have it duplicated here? can we unify `spawn_threads_and_init_comms` and `threaded_run_test`?

Reply:
We can merge by applying spawn_threads_and_init_comms  and invoking the wrapper right away. Let me fix this.